src/zcash/IncrementalMerkleTree.cpp

```diff
@@ -915,6 +915,39 @@ size_t IncrementalMerkleTree<Depth, Hash>::size() const {
     return ret;
 }
 
+template<size_t Depth, typename Hash>
+SubtreeIndex IncrementalMerkleTree<Depth, Hash>::current_subtree_index() const
+{
+    return size() >> TRACKED_SUBTREE_HEIGHT;
+}
+
+template<size_t Depth, typename Hash>
+std::optional<Hash> IncrementalMerkleTree<Depth, Hash>::complete_subtree_root() const
+{
+    auto treeSize = size();
+
+    if ((treeSize != 0) && ((treeSize % (1 << TRACKED_SUBTREE_HEIGHT)) == 0)) {
+        // We are sitting on a 2^TRACKED_SUBTREE_HEIGHT boundary here.
+        assert(left);
+        assert(right);
+        assert(parents.size() >= (TRACKED_SUBTREE_HEIGHT - 1));
+        // All of the TRACKED_SUBTREE_HEIGHT - 1 elements in the parents
+        // should be occupied or this would not be a completed subtree
+        for (size_t i = 0; i < TRACKED_SUBTREE_HEIGHT - 1; i++) {
+            assert(parents[i]);
+        }
```

Comment:
One of the important invariants to understand about the `IncrementalMerkleTree` implementation is that

* `left` is only ever unoccupied in a blank tree
* `right` can be unoccupied but if `treeSize` is a multiple of 2^k for any k, then it must be occupied
* if the tree size is nonzero, then because `treeSize` is a multiple of 2^k then it must be the case that `k - 1` of the parents are valid collapsed subtree roots

These invariants are maintained by the `append` function.

The combined check `(treeSize != 0) && ((treeSize % (1 << TRACKED_SUBTREE_HEIGHT)) == 0)` shows us that the last element appended to the tree completed a `2^TRACKED_SUBTREE_HEIGHT` subtree.

Reply:
Maybe factor this out into a `check_invariant` method?

Reply:
+1 on a factored `assert_invariant` method. I read the snippet above in meld first (without GH comments) and while the assertions help document the invariants, for ppl like me (little context) the additional description in https://github.com/zcash/zcash/pull/6691#discussion_r1213746259 would be helpful as comments in invariant assertions.

Reply:
I don't feel strongly about it.